lambda/matching.ml

```diff
@@ -1494,25 +1496,22 @@ let dbg_split_and_precompile pm next nexts =
     pretty_precompiled_res next nexts
   )
 
-let split_and_precompile_nonempty v pm =
-  let pm =
-    { pm with cases = List.map (half_simplify_nonempty ~arg:(Lvar v)) pm.cases }
-  in
-  let { me = next }, nexts = split_or (Some v) pm.cases pm.args pm.default in
-  dbg_split_and_precompile pm next nexts;
-  (next, nexts)
-
 let split_and_precompile_simplified pm =
   let { me = next }, nexts = split_no_or pm.cases pm.args pm.default [] in
   dbg_split_and_precompile pm next nexts;
   (next, nexts)
 
-let split_and_precompile ~arg_id ~arg_lambda pm =
-  let pm = { pm with cases = half_simplify_cases ~arg:arg_lambda pm.cases } in
-  let { me = next }, nexts = split_or arg_id pm.cases pm.args pm.default in
+let split_and_precompile_half_simplified ~arg pm =
+  let { me = next }, nexts = split_or arg pm.cases pm.args pm.default in
   dbg_split_and_precompile pm next nexts;
   (next, nexts)
 
+let split_and_precompile ~arg_id ~arg_lambda pm =
+  let pm =
+    { pm with cases = List.map (half_simplify_clause ~arg:arg_lambda) pm.cases }
+  in
+  split_and_precompile_half_simplified ~arg:arg_id pm
```

Comment:
This function is only used once in `do_for_multiple_match`, would it not make sense to inline it too?

Reply:
I think it make sense to have entry points for the different matrix forms, regardless of the (non-negative) number of uses. But I will try inlinining all non-simple entry points and see what happens.

Reply:
I looked at it; I think simply inlining does not improve things much, but there may be some gains to be had refactoring the `do_for_multiple_match` (I wonder if the half-simplification pass can be shared between the flattened and cannot-flatten branches). However, this would be a non-trivial change (because we need to reason about non-local exceptions), so (1) I think it should be in its own PR, rather than merged with the current "easy refactoring" one, and (2) I would rather push it at the end of the future-PR stack, rather than at the current position, so that (2a) I avoid rebase pains and (2b) I can do this refactoring in the more readable, type-informed environment.

Reply:
Refactoring `do_for_multpile_match` separately at an ulterior date sounds good to me.